### ISSUE TITLE
Add Separator at position

### ISSUE THREAD

**Issue by muichkine** [piotrmurach/tty-table#9]
Hello,

I read from the doc that it is possible to have a separator for each row:

```
table.render do |renderer|
  renderer.border.separator = :each_row
end
```

Is it possible to be able to specify where the separators apply.

For instance, being able to draw something like:

```
+---------------+----------+-----+
|white_name     |created_at|count|
+---------------+----------+-----+
|mommyville     |01/2016   |123  |
|mommyville     |02/2016   |17   |
|--------------------------------|
|alike          |01/2016   |38   |
|alike          |02/2016   |106  |
|--------------------------------|
|stephanefeugere|02/2016   |65   |
|--------------------------------|
|whysoserious   |02/2016   |126  |
|--------------------------------|
|hallcouture    |02/2016   |25   |
|--------------------------------|
|reinh          |02/2016   |3    |
|reinh          |03/2016   |1    |
+---------------+----------+-----+
```

This could be really useful also to display totals and/or subtotals rows for a table.

Cheers. TTY::Table is **SO COOL**.

**Comment by piotrmurach** [comment]
Hi Olivier,

Cheers for using the library. It still needs a lot of improvements.

Regarding your issue, I agree it would be better to provide more flexibility for separating the records. Ideally, some way to provide conditional logic. Do you have any propositions?

**Comment by muichkine** [comment]
Hello @piotrmurach ,

In my views there are two possible ways:

  1. A simple one.
     ```
     array = [["A", 1],["A", 2]]
     array.push TTY::Table.separator
     array.concat [["B", 1]]
     TTY::Table.new ["letter", "number"], array
     ```

  2. A Proc that can be passed to TTY::Table as group_by method.

```
     array = [["A", 1],["A", 2], ["B",1]]
     TTY::Table.new ["letter", "number"], array, :group_by => ->(entry){ entry.first }  # Group by letter
```

These two could also work together.  The first one is more useful when creating tables where you want to place separator "manually". For instance, draw a line before some kind of total at the bottom of a table. The second one is better when there's a simple rule to draw all separators.

That does not seem to add any weight to the TTY::Table interface. If you don't need it, don't use it (tm). :)

What do you think?

**Comment by piotrmurach** [comment]
Thanks for your feedback, it's really helpful.

As a general observation on this library and the future development, I'm wanting to keep the data and actual rendering/representation parts separate. What that means is that I wouldn't necessairly add `:separator` directly to the table data but rather instruct the renderer on how to display it. Thus you would probably handle the data on matching condtion probabably akin to:

```ruby
table.render separate_when: -> (row, col) { 
  table[row][0] != table[row-1][0]
}
```

One of the things I want this library to support is streaming of data. In which case, knowing in advance how to insert separator is impossible, you can only do it on-the-fly kind of thing. You can easily imagine a stream of tweets pouring into terminal with a watch task each getting separated by a border as they come. Having solved that problem it would lend itself very well to solving fixed size tabular data.

**Comment by muichkine** [comment]
I agree that having data separated from rendering is the best choice. Another thing in my opinion that would be real awesome is that the decision of adding a separator would not depend on the table data, but the original data the table was built with.  

One good example of this is ActiveRecord records. While you may want to print only certain attributes of the record, the decision on how to group them would depend on attributes of the record not present in the table.

I don't know how complex the TTY::Table API will look if we add all these constraints and if people really need it. (Just like I don't personally know what could be the use for data streaming :)) but I'm looking forward to talking about it. As a matter of fact I've build pretty much all my app admin interface on TTY::Table and TTY::Prompt so I'm a big user. :)

To give you perspective, there are two reasons to that. One, it's faster than doing an admin web page. Two, I'm entertaining the idea of adding to my app a bot interface (a slack bot, irc bot, whatever bot) and I think TTY::Prompt/Table is a great proxy to the use case.

**Comment by piotrmurach** [comment]
A big user - fantastic! I'm glad the tty components are helping you build terminal interface. I sense a future collaboration on this front if you are interested as I have some plans that may be exciting to yourself. I probably reserve some time for the idea to fully develop in my brain before I share it with yourself.

Regarding streaming of data, I suggest having a look at [this issue](https://github.com/piotrmurach/tty-table/issues/1) where I have put my initial thoughts. I haven't had the need for this feature hence I have never implemented this functionality but I can totally see the value of it and plan to add it in future releases.

**Comment by muichkine** [comment]
Hello @piotrmurach, do you have an example of usage for streaming tables?

**Comment by wesm-outreach** [comment]
I'm also very interested in this feature.

**Comment by ghostsquad** [comment]
I'm also very interested in this feature.

**Comment by piotrmurach** [comment]
I've been recently really focused on [tty-prompt](https://github.com/piotrmurach/tty-prompt) and providing good Windows support. There are still few features to implement such as `autocomplete` that people were asking for a long time. Unfortunately I do this in my spare time and would love to attend immediately to all issues but I simply cannot. This library needs more love and I will provide it but I doubt I will have time this month, just trying to be realistic. Having said that, if I receive PR for this feature I promise to review it quickly.

**Comment by ghostsquad** [comment]
I'd love to contribute back to this project. If I have time, I'll submit a PR.

**Comment by piotrmurach** [comment]
This is implemented in the latest `v0.11.0` release. See [separator section](https://github.com/piotrmurach/tty-table#353-separator)
